### PR TITLE
In-app image viewer

### LINKED ISSUE #397: Allow images in posts to be expanded and zoomed

### PR DESCRIPTION
**Pull request by pipe01**
Whenever a post's image is clicked on the home page, the full image will be opened in the image viewer, which has the usual pinch zoom and pan functionality. There's also a save button in the top right corner which will download the image to the device's pictures folder, and a back button on the top left.

Everything is working already, I'm just waiting for https://github.com/usuiat/Zoomable to release a new version which will include a couple features that I need.

Fixes #397

### CODE REVIEW

**Comment by pipe01** [comment]
Oops yeah sorry, the download button is on the top right, not top left. On the top left there's a back button as you said.

**Comment by TrollBlox** [comment]
I can't wait to see this in the app! Is all that's left waiting for a new version of Zoomable?

**Comment by pipe01** [comment]
Yup, and also any feedback @dessalines may have.

File: app/src/main/java/com/jerboa/ui/components/person/PersonProfile.kt [hunk 1]
```diff
@@ -58,12 +69,21 @@ fun PersonProfileTopSection(
...
+                        },
                 )
             }
             Box(modifier = Modifier.padding(MEDIUM_PADDING)) {
                 personView.person.avatar?.also {
-                    LargerCircularIcon(icon = it)
+                    LargerCircularIcon(
+                        icon = it,
+                        modifier = Modifier.clickable {
+                            showImage = personView.person.avatar
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/ui/components/person/PersonProfile.kt:84]: Nice, I like this construction.

File: app/src/main/java/com/jerboa/ui/components/common/ImageViewerDialog.kt [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+suspend fun SaveImage(url: String, context: Context) {
+    val request = ImageRequest.Builder(context)
+        .data(url)
+        .crossfade(true)
+        .target(
+            onSuccess = {
+                val fileName = Uri.parse(url).pathSegments.last()
+                saveBitmap(context, it.toBitmap(), Bitmap.CompressFormat.WEBP, "image/webp", fileName)
```
Comment (dessalines) [inline app/src/main/java/com/jerboa/ui/components/common/ImageViewerDialog.kt:136]: I've gone back and forth on webp... it is smaller and that's nice, but also according to the dev of a picture library I really like (photoprism), google's support is terrible and nearly all open-source devs that work with picture software hate it. Plus people will probably want to download the high-res originals, in whatever format the pictrs server provides.

I'm fairly sure the imageUrl I'm providing can be jpg, jpeg, png, or webp for the thumbnails only, so I spose just match whatever's there.

Comment (pipe01) [reply app/src/main/java/com/jerboa/ui/components/common/ImageViewerDialog.kt:138]: Yeah it's not that great in terms of support, even in Go which is made by Google the support is a bit iffy. What I wanted to do initially was to just download the image straight to a file, but it seems like there's no way to get the bitmap's raw data. I'll see what I can do.

File: app/src/main/java/com/jerboa/ui/components/common/ImageViewerDialog.kt [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import androidx.core.graphics.drawable.toBitmap
+import coil.compose.rememberAsyncImagePainter
+import coil.imageLoader
+import coil.request.ImageRequest
+import com.jerboa.saveBitmap
+import kotlinx.coroutines.launch
+import net.engawapg.lib.zoomable.rememberZoomState
+import net.engawapg.lib.zoomable.zoomable
+
+val backColor = Color.Black
```
[the remark this answers is not in the document]
Comment (pipe01) [reply app/src/main/java/com/jerboa/ui/components/common/ImageViewerDialog.kt]: Seems like the scrim color was made for this, I'll use that instead.

**Comment by pipe01** [comment]
I gonna try to add download notifications, and then this can be merged if there are no further issues.

**Review by twizmwazin — APPROVED** [review]
I tested it out and seems to work well. It doesn't appear to work for images embedded in comments yet.

**Review by dessalines — APPROVED** [review]

**Comment by shombando** [comment]
@pipe01 thanks for all your great functionality that got merged to 0.33-alpha, the experience is improving in subtle but important ways!

**Pull request merged**